pkg/sql/catalog/internal/catval/validate.go

```diff
@@ -0,0 +1,448 @@
+// specified level.
+// Some of these checks may require cross-referencing with other descriptors,
+// hence the need for a ctx and a ValidationDereferencer.
+// If one of these checks surfaces an error, that error is wrapped with a prefix
+// identifying the descriptor being validated.
+// Validate returns a ValidationErrors interface which can provide the errors
+// either as a slice or combined as one.
+func Validate(
+	ctx context.Context,
+	vd ValidationDereferencer,
+	telemetry catalog.ValidationTelemetry,
+	targetLevel catalog.ValidationLevel,
+	descriptors ...catalog.Descriptor,
+) catalog.ValidationErrors {
+	vea := validationErrorAccumulator{
+		ValidationTelemetry: telemetry,
+		targetLevel:         targetLevel,
+	}
+	// Internal descriptor consistency checks.
+	if !vea.validateDescriptorsAtLevel(
+		catalog.ValidationLevelSelfOnly,
+		descriptors,
+		func(desc catalog.Descriptor) {
+			desc.ValidateSelf(&vea)
+		}) {
+		return vea.errors
+	}
+	// Collect descriptors referenced by the validated descriptors.
+	// These are their immediate neighbors in the reference graph, and in some
+	// special cases those neighbors' immediate neighbors also.
+	vdg, descGetterErr := collectDescriptorsForValidation(ctx, vd, descriptors)
+	if descGetterErr != nil {
+		vea.reportDescGetterError(collectingReferencedDescriptors, descGetterErr)
+		return vea.errors
+	}
+	// Descriptor cross-reference checks.
+	if !vea.validateDescriptorsAtLevel(
+		catalog.ValidationLevelCrossReferences,
+		descriptors,
+		func(desc catalog.Descriptor) {
+			if !desc.Dropped() {
+				desc.ValidateCrossReferences(&vea, vdg)
+			}
+		}) {
+		return vea.errors
+	}
+	// Collect descriptor namespace table entries, if running namespace checks.
+	if catalog.ValidationLevelNamespace&targetLevel != 0 {
+		descGetterErr = vdg.addNamespaceEntries(ctx, descriptors, vd)
+		if descGetterErr != nil {
+			vea.reportDescGetterError(collectingNamespaceEntries, descGetterErr)
+			return vea.errors
+		}
+	}
+	// Namespace validation checks
+	if !vea.validateDescriptorsAtLevel(
+		catalog.ValidationLevelNamespace,
+		descriptors,
+		func(desc catalog.Descriptor) {
+			validateNamespace(desc, &vea, vdg.namespace)
+		}) {
+		return vea.errors
+	}
+	// Descriptor pre-txn-commit checks.
+	_ = vea.validateDescriptorsAtLevel(
+		catalog.ValidationLevelAllPreTxnCommit,
+		descriptors,
+		func(desc catalog.Descriptor) {
+			if !desc.Dropped() {
+				desc.ValidateTxnCommit(&vea, vdg)
+			}
+		})
+	return vea.errors
+}
+
+// ValidationDereferencer is an interface to retrieve descriptors and namespace
+// entries.
+// Lookups are performed on a best-effort basis. When a descriptor or namespace
+// entry cannot be found, the zero-value is returned, with no error.
+type ValidationDereferencer interface {
+	DereferenceDescriptors(ctx context.Context, reqs []descpb.ID) ([]catalog.Descriptor, error)
+	DereferenceDescriptorIDs(ctx context.Context, requests []descpb.NameInfo) ([]descpb.ID, error)
+}
```

Comment:
this is all that remains of the catalog.DescGetter and it's implemented only by nstree.Catalog and descs.Collection and some placeholder in catkv which is the former oneLevelUncachedDescGetter. Long-term goal is to replace the latter with descs.Collection but this might alter behaviour so I'm putting it off.